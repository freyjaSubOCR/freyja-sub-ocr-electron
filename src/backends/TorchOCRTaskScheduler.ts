import TorchOCR from './TorchOCR'
import { ipcMain } from 'electron'
import logger from '@/logger'
import { Tensor } from 'torch-js'
import Config from '@/config'
import { SubtitleInfo, VideoProperties } from '@/interfaces'

class TorchOCRTaskScheduler {
    private torchOCR: TorchOCR = new TorchOCR()
    private currentProcessingFrame = 0
    private subtitleInfos: SubtitleInfo[] = []

    registerIPCListener(): void {
        ipcMain.handle('TorchOCRTaskScheduler:Init', async (e, ...args) => {
            try {
                return await this.Init(args[0])
            } catch (error) {
                logger.error(error.message)
                return null
            }
        })
    }

    async Init(path: string): Promise<void> {
        this.torchOCR.InitRCNN()
        await this.torchOCR.InitOCR()
        await this.torchOCR.InitVideoPlayer(path)
    }

    async Start(): Promise<SubtitleInfo[]> {
        let step = Config.BatchSize
        let tensorDataPromise = new Promise(resolve => resolve())
        let rcnnPromise = new Promise(resolve => resolve())
        let ocrPromise = new Promise(resolve => resolve())
        const ocrPromiseBuffer = [ocrPromise, ocrPromise, ocrPromise, ocrPromise]
        if (this.torchOCR.videoProperties === undefined) {
            throw new Error('VideoPlayer is not initialized')
        }

        for (let frame = 0; frame <= this.torchOCR.videoProperties.lastFrame; frame += step) {
            if (this.torchOCR.videoProperties.lastFrame + 1 - frame < step) {
                step = this.torchOCR.videoProperties.lastFrame + 1 - frame
            }

            tensorDataPromise = Promise.all([tensorDataPromise, ocrPromiseBuffer[0]]).then(async () => {
                const rawImg: Buffer[] = []
                for (const i of Array(step).keys()) {
                    console.log(`read frame ${i + frame}`)
                    console.log(`last frame ${(this.torchOCR.videoProperties as VideoProperties).lastFrame}`)
                    rawImg.push(await this.torchOCR.ReadRawFrame(i + frame))
                }
                this.currentProcessingFrame = frame
                const inputTensor = this.torchOCR.BufferToImgTensor(rawImg, 600)
                return inputTensor
            })

            rcnnPromise = Promise.all([rcnnPromise, tensorDataPromise]).then(async (values) => {
                const inputTensor = values[1] as Tensor
                const rcnnResults = await this.torchOCR.RCNNForward(inputTensor)
                return rcnnResults
            })

            ocrPromise = Promise.all([ocrPromise, tensorDataPromise, rcnnPromise]).then(async (values) => {
                const inputTensor = values[1] as Tensor
                const rcnnResults = values[2] as Record<string, Tensor>[]
                const subtitleInfos = this.torchOCR.RCNNParse(rcnnResults)
                if (subtitleInfos.length !== 0) {
                    const boxesTensor = this.torchOCR.SubtitleInfoToTensor(subtitleInfos)
                    const ocrResults = this.torchOCR.OCRParse(await this.torchOCR.OCRForward(inputTensor, boxesTensor))

                    for (const i of subtitleInfos.keys()) {
                        subtitleInfos[i].text = ocrResults[i]
                        subtitleInfos[i].startFrame = subtitleInfos[i].startFrame + frame
                        subtitleInfos[i].endFrame = subtitleInfos[i].endFrame + frame
                        this.subtitleInfos.push(subtitleInfos[i])
                    }
                    boxesTensor.free()
                }
                inputTensor.free()
            })
            ocrPromiseBuffer.shift()
            ocrPromiseBuffer.push(ocrPromise)
        }
        await Promise.all([tensorDataPromise, rcnnPromise, ocrPromise])
        return this.subtitleInfos
    }
}

export default TorchOCRTaskScheduler
