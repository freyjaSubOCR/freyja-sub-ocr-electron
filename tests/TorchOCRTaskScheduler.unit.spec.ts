import TorchOCRTaskScheduler from '@/backends/TorchOCRTaskScheduler'

describe('TorchOCRTaskScheduler.ts', () => {
    it('Intergration test', async () => {
        const task = new TorchOCRTaskScheduler()
        await task.Init('tests/files/sample.mp4')
        const result = await task.Start()
        expect(result).toMatchSnapshot()
    }, 100000)
})
