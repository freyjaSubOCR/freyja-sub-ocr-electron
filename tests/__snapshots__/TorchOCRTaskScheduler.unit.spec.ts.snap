// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TorchOCRTaskScheduler.ts Intergration test 1`] = `
Array [
  Object {
    "box": Int32Array [
      774,
      361,
      1143,
      462,
    ],
    "endFrame": 3,
    "startFrame": 2,
    "text": "我是神野惠",
  },
  Object {
    "box": Int32Array [
      774,
      361,
      1143,
      462,
    ],
    "endFrame": 4,
    "startFrame": 3,
    "text": "我是神野惠",
  },
  Object {
    "box": Int32Array [
      769,
      362,
      1142,
      461,
    ],
    "endFrame": 21,
    "startFrame": 20,
    "text": " 我是神野惠",
  },
  Object {
    "box": Int32Array [
      769,
      362,
      1142,
      461,
    ],
    "endFrame": 22,
    "startFrame": 21,
    "text": " 我是神野惠",
  },
  Object {
    "box": Int32Array [
      769,
      362,
      1142,
      461,
    ],
    "endFrame": 23,
    "startFrame": 22,
    "text": " 我是神野惠",
  },
  Object {
    "box": Int32Array [
      769,
      362,
      1142,
      461,
    ],
    "endFrame": 24,
    "startFrame": 23,
    "text": " 我是神野惠",
  },
  Object {
    "box": Int32Array [
      609,
      362,
      1314,
      465,
    ],
    "endFrame": 41,
    "startFrame": 40,
    "text": "是和泉纱雾的同班同学",
  },
  Object {
    "box": Int32Array [
      609,
      362,
      1314,
      465,
    ],
    "endFrame": 42,
    "startFrame": 41,
    "text": "是和泉纱雾的同班同学",
  },
  Object {
    "box": Int32Array [
      609,
      362,
      1314,
      465,
    ],
    "endFrame": 43,
    "startFrame": 42,
    "text": "是和泉纱雾的同班同学",
  },
  Object {
    "box": Int32Array [
      609,
      362,
      1314,
      465,
    ],
    "endFrame": 44,
    "startFrame": 43,
    "text": "是和泉纱雾的同班同学",
  },
  Object {
    "box": Int32Array [
      607,
      362,
      1313,
      462,
    ],
    "endFrame": 61,
    "startFrame": 60,
    "text": "是和泉纱雾的同班同学",
  },
  Object {
    "box": Int32Array [
      607,
      362,
      1313,
      462,
    ],
    "endFrame": 62,
    "startFrame": 61,
    "text": "是和泉纱雾的同班同学",
  },
  Object {
    "box": Int32Array [
      608,
      362,
      1313,
      462,
    ],
    "endFrame": 63,
    "startFrame": 62,
    "text": "是和泉纱雾的同班同学",
  },
  Object {
    "box": Int32Array [
      608,
      362,
      1313,
      462,
    ],
    "endFrame": 64,
    "startFrame": 63,
    "text": "是和泉纱雾的同班同学",
  },
  Object {
    "box": Int32Array [
      608,
      362,
      1312,
      462,
    ],
    "endFrame": 81,
    "startFrame": 80,
    "text": "是和泉纱雾的同班同学",
  },
  Object {
    "box": Int32Array [
      608,
      362,
      1312,
      462,
    ],
    "endFrame": 82,
    "startFrame": 81,
    "text": "是和泉纱雾的同班同学",
  },
  Object {
    "box": Int32Array [
      537,
      361,
      1378,
      460,
    ],
    "endFrame": 101,
    "startFrame": 100,
    "text": "你为什么要b她进我们家啊",
  },
  Object {
    "box": Int32Array [
      537,
      361,
      1378,
      460,
    ],
    "endFrame": 102,
    "startFrame": 101,
    "text": "你为什么要b她进我们家啊",
  },
  Object {
    "box": Int32Array [
      537,
      361,
      1378,
      460,
    ],
    "endFrame": 103,
    "startFrame": 102,
    "text": "你为什么要b她进我们家啊",
  },
  Object {
    "box": Int32Array [
      534,
      361,
      1378,
      460,
    ],
    "endFrame": 104,
    "startFrame": 103,
    "text": "你为什么要b她进我家啊",
  },
  Object {
    "box": Int32Array [
      534,
      361,
      1378,
      460,
    ],
    "endFrame": 121,
    "startFrame": 120,
    "text": "你为什么要b她进我家啊",
  },
  Object {
    "box": Int32Array [
      534,
      361,
      1378,
      460,
    ],
    "endFrame": 122,
    "startFrame": 121,
    "text": "你为什么要b她进我家啊",
  },
  Object {
    "box": Int32Array [
      539,
      361,
      1382,
      460,
    ],
    "endFrame": 123,
    "startFrame": 122,
    "text": "你为什么要凶她进我1家啊",
  },
  Object {
    "box": Int32Array [
      539,
      361,
      1382,
      460,
    ],
    "endFrame": 124,
    "startFrame": 123,
    "text": "你为什么要凶她进我1家啊",
  },
  Object {
    "box": Int32Array [
      539,
      361,
      1382,
      460,
    ],
    "endFrame": 141,
    "startFrame": 140,
    "text": "你为什么要凶她进我1家啊",
  },
  Object {
    "box": Int32Array [
      539,
      361,
      1382,
      460,
    ],
    "endFrame": 142,
    "startFrame": 141,
    "text": "你为什么要凶她进我1家啊",
  },
  Object {
    "box": Int32Array [
      338,
      361,
      1586,
      461,
    ],
    "endFrame": 161,
    "startFrame": 160,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡了",
  },
  Object {
    "box": Int32Array [
      337,
      361,
      1587,
      462,
    ],
    "endFrame": 162,
    "startFrame": 161,
    "text": "跟我差不多岁数 的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      337,
      361,
      1587,
      462,
    ],
    "endFrame": 163,
    "startFrame": 162,
    "text": "跟我差不多岁数 的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      337,
      361,
      1587,
      462,
    ],
    "endFrame": 164,
    "startFrame": 163,
    "text": "跟我差不多岁数 的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      337,
      362,
      1584,
      461,
    ],
    "endFrame": 181,
    "startFrame": 180,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      337,
      362,
      1584,
      461,
    ],
    "endFrame": 182,
    "startFrame": 181,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      337,
      362,
      1584,
      461,
    ],
    "endFrame": 183,
    "startFrame": 182,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      337,
      362,
      1584,
      461,
    ],
    "endFrame": 184,
    "startFrame": 183,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 201,
    "startFrame": 200,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 202,
    "startFrame": 201,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 203,
    "startFrame": 202,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 204,
    "startFrame": 203,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 221,
    "startFrame": 220,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 222,
    "startFrame": 221,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 223,
    "startFrame": 222,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 224,
    "startFrame": 223,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 241,
    "startFrame": 240,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 242,
    "startFrame": 241,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      329,
      362,
      1584,
      460,
    ],
    "endFrame": 243,
    "startFrame": 242,
    "text": "跟我差不多岁数的女孩子都最喜欢鸡鸡了",
  },
  Object {
    "box": Int32Array [
      742,
      366,
      1145,
      458,
    ],
    "endFrame": 261,
    "startFrame": 260,
    "text": "最…最喜欢？",
  },
  Object {
    "box": Int32Array [
      742,
      366,
      1145,
      459,
    ],
    "endFrame": 262,
    "startFrame": 261,
    "text": "最…最喜欢？",
  },
  Object {
    "box": Int32Array [
      742,
      366,
      1145,
      459,
    ],
    "endFrame": 263,
    "startFrame": 262,
    "text": "最…最喜欢？",
  },
  Object {
    "box": Int32Array [
      742,
      366,
      1145,
      459,
    ],
    "endFrame": 264,
    "startFrame": 263,
    "text": "最…最喜欢？",
  },
  Object {
    "box": Int32Array [
      262,
      361,
      1666,
      464,
    ],
    "endFrame": 301,
    "startFrame": 300,
    "text": "名字m「把小和泉从房间里拖出来lJ同盟",
  },
  Object {
    "box": Int32Array [
      262,
      361,
      1666,
      464,
    ],
    "endFrame": 302,
    "startFrame": 301,
    "text": "名字m「把小和泉从房间里拖出来lJ同盟",
  },
  Object {
    "box": Int32Array [
      262,
      361,
      1666,
      464,
    ],
    "endFrame": 303,
    "startFrame": 302,
    "text": "名字m「把小和泉从房间里拖出来lJ同盟",
  },
  Object {
    "box": Int32Array [
      262,
      361,
      1666,
      464,
    ],
    "endFrame": 304,
    "startFrame": 303,
    "text": "名字m「把小和泉从房间里拖出来lJ同盟",
  },
  Object {
    "box": Int32Array [
      262,
      360,
      1666,
      464,
    ],
    "endFrame": 321,
    "startFrame": 320,
    "text": "名字 把xk和最从房里抽出汝！」同盟",
  },
  Object {
    "box": Int32Array [
      262,
      360,
      1666,
      464,
    ],
    "endFrame": 322,
    "startFrame": 321,
    "text": "名字 把xk和最从房里抽出汝！」同盟",
  },
  Object {
    "box": Int32Array [
      262,
      361,
      1666,
      464,
    ],
    "endFrame": 323,
    "startFrame": 322,
    "text": "名字m「把小和泉从房间里拖出来lJ同盟",
  },
  Object {
    "box": Int32Array [
      262,
      361,
      1666,
      464,
    ],
    "endFrame": 324,
    "startFrame": 323,
    "text": "名字m「把小和泉从房间里拖出来lJ同盟",
  },
  Object {
    "box": Int32Array [
      262,
      361,
      1666,
      464,
    ],
    "endFrame": 341,
    "startFrame": 340,
    "text": "名字m「把小和泉从房间里拖出来lJ同盟",
  },
  Object {
    "box": Int32Array [
      262,
      361,
      1666,
      464,
    ],
    "endFrame": 342,
    "startFrame": 341,
    "text": "名字m「把小和泉从房间里拖出来lJ同盟",
  },
  Object {
    "box": Int32Array [
      262,
      360,
      1666,
      464,
    ],
    "endFrame": 343,
    "startFrame": 342,
    "text": "名字 把xk和最从房里抽出汝！」同盟",
  },
  Object {
    "box": Int32Array [
      262,
      361,
      1665,
      464,
    ],
    "endFrame": 344,
    "startFrame": 343,
    "text": "名字m「把小和泉从房间里拖出来lJ同盟",
  },
  Object {
    "box": Int32Array [
      269,
      360,
      1657,
      460,
    ],
    "endFrame": 361,
    "startFrame": 360,
    "text": "名字就叫「把小和泉从房间里拖出求！」同盟",
  },
  Object {
    "box": Int32Array [
      270,
      361,
      1655,
      460,
    ],
    "endFrame": 362,
    "startFrame": 361,
    "text": "名字就叫「把k小和泉从房间里拖出来！」同盟",
  },
  Object {
    "box": Int32Array [
      269,
      361,
      1655,
      460,
    ],
    "endFrame": 363,
    "startFrame": 362,
    "text": "名字就叫「把小和泉从房间里拖出来！」同盟",
  },
  Object {
    "box": Int32Array [
      268,
      361,
      1657,
      460,
    ],
    "endFrame": 364,
    "startFrame": 363,
    "text": "名字就叫「把小和泉从房间里拖出来l！」同盟",
  },
  Object {
    "box": Int32Array [
      738,
      362,
      1186,
      458,
    ],
    "endFrame": 381,
    "startFrame": 380,
    "text": "不出去 不出去",
  },
  Object {
    "box": Int32Array [
      740,
      363,
      1186,
      459,
    ],
    "endFrame": 382,
    "startFrame": 381,
    "text": "不出去 不出去",
  },
  Object {
    "box": Int32Array [
      740,
      363,
      1186,
      459,
    ],
    "endFrame": 383,
    "startFrame": 382,
    "text": "不出去 不出去",
  },
  Object {
    "box": Int32Array [
      740,
      363,
      1186,
      459,
    ],
    "endFrame": 384,
    "startFrame": 383,
    "text": "不出去 不出去",
  },
  Object {
    "box": Int32Array [
      741,
      363,
      1186,
      458,
    ],
    "endFrame": 401,
    "startFrame": 400,
    "text": "不出去 不出去",
  },
  Object {
    "box": Int32Array [
      741,
      363,
      1186,
      458,
    ],
    "endFrame": 402,
    "startFrame": 401,
    "text": "不出去 不出去",
  },
  Object {
    "box": Int32Array [
      738,
      362,
      1186,
      458,
    ],
    "endFrame": 403,
    "startFrame": 402,
    "text": "不出去 不出去",
  },
  Object {
    "box": Int32Array [
      608,
      362,
      1312,
      458,
    ],
    "endFrame": 404,
    "startFrame": 403,
    "text": "我说不出去就是不出去",
  },
  Object {
    "box": Int32Array [
      608,
      362,
      1312,
      458,
    ],
    "endFrame": 421,
    "startFrame": 420,
    "text": "我说不出去就是不出去",
  },
  Object {
    "box": Int32Array [
      608,
      362,
      1312,
      458,
    ],
    "endFrame": 422,
    "startFrame": 421,
    "text": "我说不出去就是不出去",
  },
  Object {
    "box": Int32Array [
      608,
      362,
      1312,
      458,
    ],
    "endFrame": 423,
    "startFrame": 422,
    "text": "我说不出去就是不出去",
  },
  Object {
    "box": Int32Array [
      606,
      362,
      1312,
      458,
    ],
    "endFrame": 424,
    "startFrame": 423,
    "text": "我说不出去就是不出去",
  },
  Object {
    "box": Int32Array [
      536,
      362,
      1349,
      460,
    ],
    "endFrame": 442,
    "startFrame": 441,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      535,
      362,
      1349,
      460,
    ],
    "endFrame": 443,
    "startFrame": 442,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      535,
      362,
      1349,
      460,
    ],
    "endFrame": 444,
    "startFrame": 443,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      536,
      361,
      1348,
      460,
    ],
    "endFrame": 461,
    "startFrame": 460,
    "text": "哥哥想tb妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      536,
      361,
      1348,
      460,
    ],
    "endFrame": 462,
    "startFrame": 461,
    "text": "哥哥想tb妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      535,
      361,
      1348,
      460,
    ],
    "endFrame": 463,
    "startFrame": 462,
    "text": "哥哥想tb妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      540,
      362,
      1347,
      459,
    ],
    "endFrame": 464,
    "startFrame": 463,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      540,
      362,
      1347,
      459,
    ],
    "endFrame": 481,
    "startFrame": 480,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      541,
      362,
      1348,
      460,
    ],
    "endFrame": 482,
    "startFrame": 481,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      541,
      362,
      1348,
      460,
    ],
    "endFrame": 483,
    "startFrame": 482,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      540,
      362,
      1348,
      459,
    ],
    "endFrame": 484,
    "startFrame": 483,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      538,
      360,
      1345,
      459,
    ],
    "endFrame": 501,
    "startFrame": 500,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      538,
      360,
      1346,
      459,
    ],
    "endFrame": 502,
    "startFrame": 501,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      536,
      360,
      1346,
      459,
    ],
    "endFrame": 503,
    "startFrame": 502,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      537,
      361,
      1347,
      459,
    ],
    "endFrame": 504,
    "startFrame": 503,
    "text": "哥哥想凶妹妹变成什么样？",
  },
  Object {
    "box": Int32Array [
      773,
      361,
      1108,
      461,
    ],
    "endFrame": 541,
    "startFrame": 540,
    "text": "和泉政宗?？",
  },
  Object {
    "box": Int32Array [
      773,
      361,
      1108,
      461,
    ],
    "endFrame": 542,
    "startFrame": 541,
    "text": "和泉政宗?？",
  },
  Object {
    "box": Int32Array [
      773,
      361,
      1108,
      461,
    ],
    "endFrame": 543,
    "startFrame": 542,
    "text": "和泉政宗?？",
  },
  Object {
    "box": Int32Array [
      776,
      361,
      1108,
      461,
    ],
    "endFrame": 544,
    "startFrame": 543,
    "text": "和泉政宗?？",
  },
  Object {
    "box": Int32Array [
      774,
      361,
      1107,
      461,
    ],
    "endFrame": 561,
    "startFrame": 560,
    "text": "和泉政宗？",
  },
  Object {
    "box": Int32Array [
      535,
      364,
      1352,
      460,
    ],
    "endFrame": 581,
    "startFrame": 580,
    "text": "居然是如此年轻的女孩子？",
  },
  Object {
    "box": Int32Array [
      534,
      363,
      1349,
      461,
    ],
    "endFrame": 582,
    "startFrame": 581,
    "text": "居然是如此年轻的女孩子？",
  },
  Object {
    "box": Int32Array [
      534,
      362,
      1350,
      461,
    ],
    "endFrame": 583,
    "startFrame": 582,
    "text": "居然是如此年轻的女孩子？",
  },
  Object {
    "box": Int32Array [
      544,
      362,
      1348,
      460,
    ],
    "endFrame": 584,
    "startFrame": 583,
    "text": "居然是如此年轻的女孩？",
  },
  Object {
    "box": Int32Array [
      542,
      361,
      1351,
      463,
    ],
    "endFrame": 601,
    "startFrame": 600,
    "text": "居然是如此年轻的女孩子？",
  },
  Object {
    "box": Int32Array [
      536,
      362,
      1350,
      462,
    ],
    "endFrame": 602,
    "startFrame": 601,
    "text": "居然是如此年轻的女孩子？",
  },
  Object {
    "box": Int32Array [
      535,
      362,
      1349,
      462,
    ],
    "endFrame": 603,
    "startFrame": 602,
    "text": "居然是如此年轻的女孩子？",
  },
  Object {
    "box": Int32Array [
      539,
      362,
      1347,
      462,
    ],
    "endFrame": 604,
    "startFrame": 603,
    "text": "居然是如此年轻的女孩子？",
  },
  Object {
    "box": Int32Array [
      233,
      362,
      1706,
      460,
    ],
    "endFrame": 641,
    "startFrame": 640,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      233,
      362,
      1706,
      460,
    ],
    "endFrame": 642,
    "startFrame": 641,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      233,
      362,
      1706,
      460,
    ],
    "endFrame": 643,
    "startFrame": 642,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      233,
      362,
      1706,
      460,
    ],
    "endFrame": 644,
    "startFrame": 643,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      233,
      362,
      1706,
      460,
    ],
    "endFrame": 661,
    "startFrame": 660,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      233,
      362,
      1706,
      460,
    ],
    "endFrame": 662,
    "startFrame": 661,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      233,
      362,
      1706,
      460,
    ],
    "endFrame": 663,
    "startFrame": 662,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      233,
      362,
      1706,
      460,
    ],
    "endFrame": 664,
    "startFrame": 663,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      227,
      362,
      1706,
      461,
    ],
    "endFrame": 681,
    "startFrame": 680,
    "text": " 我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      217,
      362,
      1706,
      461,
    ],
    "endFrame": 682,
    "startFrame": 681,
    "text": " 我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      220,
      362,
      1705,
      460,
    ],
    "endFrame": 683,
    "startFrame": 682,
    "text": " 我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      225,
      362,
      1703,
      460,
    ],
    "endFrame": 684,
    "startFrame": 683,
    "text": " 我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      234,
      362,
      1715,
      461,
    ],
    "endFrame": 701,
    "startFrame": 700,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      234,
      362,
      1715,
      461,
    ],
    "endFrame": 702,
    "startFrame": 701,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      235,
      362,
      1702,
      461,
    ],
    "endFrame": 703,
    "startFrame": 702,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      235,
      362,
      1702,
      461,
    ],
    "endFrame": 704,
    "startFrame": 703,
    "text": "我的下一个作品要请情色漫画老师来担任插画师",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 741,
    "startFrame": 740,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 742,
    "startFrame": 741,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 743,
    "startFrame": 742,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 744,
    "startFrame": 743,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      573,
      31,
      1443,
      123,
    ],
    "endFrame": 761,
    "startFrame": 760,
    "text": "耻 …与  人 t:",
  },
  Object {
    "box": Int32Array [
      573,
      31,
      1443,
      123,
    ],
    "endFrame": 762,
    "startFrame": 761,
    "text": "耻 …与  人 t:",
  },
  Object {
    "box": Int32Array [
      573,
      31,
      1443,
      123,
    ],
    "endFrame": 763,
    "startFrame": 762,
    "text": "耻 …与  人 t:",
  },
  Object {
    "box": Int32Array [
      573,
      31,
      1443,
      123,
    ],
    "endFrame": 764,
    "startFrame": 763,
    "text": "耻 …与  人 t:",
  },
  Object {
    "box": Int32Array [
      573,
      31,
      1443,
      123,
    ],
    "endFrame": 781,
    "startFrame": 780,
    "text": "耻 …与  人 t:",
  },
  Object {
    "box": Int32Array [
      573,
      31,
      1443,
      123,
    ],
    "endFrame": 782,
    "startFrame": 781,
    "text": "耻 …与  人 t:",
  },
  Object {
    "box": Int32Array [
      573,
      31,
      1443,
      123,
    ],
    "endFrame": 783,
    "startFrame": 782,
    "text": "耻 …与  人 t:",
  },
  Object {
    "box": Int32Array [
      573,
      31,
      1443,
      123,
    ],
    "endFrame": 784,
    "startFrame": 783,
    "text": "耻 …与  人 t:",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 801,
    "startFrame": 800,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 802,
    "startFrame": 801,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 803,
    "startFrame": 802,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 804,
    "startFrame": 803,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 821,
    "startFrame": 820,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 822,
    "startFrame": 821,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 823,
    "startFrame": 822,
    "text": "？ 乃…与  人 t",
  },
  Object {
    "box": Int32Array [
      572,
      31,
      1443,
      123,
    ],
    "endFrame": 824,
    "startFrame": 823,
    "text": "？ 乃…与  人 t",
  },
]
`;
